### ISSUE TITLE
Feature Proposal: Audit Logging

### ISSUE THREAD

**Issue by mbramson** [tompave/fun_with_flags#59]
It would be great to have the ability to track who toggled what features and to what. This can be useful both from a troubleshooting standpoint but also to satisfy auditing/compliance requirements. 

It would also be lovely if this worked with https://github.com/tompave/fun_with_flags_ui, the more seamless the better. As such this proposal encompasses both libraries, although the changes to `FunWithFlags` do stand on their own.

## `FunWithFlags` changes.

I'd like to add a new function: 
`FunWithFlags.add_audit_log(action, flag_name, current_user, opts)`

It would be called like so:
```
FunWithFlags.add_audit_log(:enable, :my_feature, 1234, [for_percentage_of: {:time, ratio}])
```

Under the hood this would call a new callback on the `Persistent` behavior:

`c:FunWithFlags.Store.Persistent.add_audit_log/4` which we'd implement for both
Redix and Ecto.

These function implementations would simply persist an audit log to either a new redis key or a new postgres table. This would require updating the readme to include this new table in a migration (shouldn't be difficult to make this optional).

Additionally, it could be useful to also add another function to the `Persistent` behavior which wraps other things in a transaction. This would be useful in `fun_with_flags_ui` portion of these changes to ensure that we don't have situations where an audit log gets added but the feature flag doesn't change and vice versa. Thinking a function signature would look something like:

`c:FunWithFlags.Store.Persistent.transaction/1` which would just call `Ecto.Transaction` or `Redix.transaction_pipeline/3` essentially.

All of the above changes I believe can stand alone as useful features to `FunWithFlags`.

## `FunWithFlags.UI` changes.

There are a number of places in `FunWithFlagsUI` that call `FunWithFlags.enable/2`,`FunWithFlags.disable/2`, and `FunWithFlags.clear/2` directly.

The idea is to wrap those calls in new functions defined in `FunWithFlags.UI.Utils` that call out to the previously mentioned `FunWithFlags.Store.Persistent.transaction/1` and then calls `FunWithFlags.enable/2`,`FunWithFlags.disable/2`, or `FunWithFlags.clear/2`, as well as `FunWithFlags.add_audit_log/4`.

The one missing piece at this point is the method of retrieving the current user.

Would love feedback on this next part:

We could make some assumptions or have some opinions.

We could enforce that the current user is always in `conn.assigns.fun_with_flags_current_user` and leave it up to the users of `FunWithFlags.UI` to add a plug which puts the current user there.

Or we could allow users to define a function that accepts the `%Plug.Conn{}` struct as an argument and spits out the current user. This is way more flexible, and doesn't force users to add something to `assigns`, but it's slightly more elaborate to configure. You'd have to pass a reference to this function into application config.

We could also go even further and define a behavior with a callback that accepts a `%Plug.Conn{}` struct and outputs the current user, and then force users to implement that callback on a module that they can then pass in as config. This is also a bit elaborate.

Regardless of which one we go with, we'd want to configure this behavior in application config and ensure that if no `:audit_logging` config is supplied (should be the state of all things that have implemented this to date), then things simply work and no audit logging is attempted. This should be an opt-in feature.

## Audit Logging UI

The above is all we need to get audit logging implemented and working fairly seemlessly, but we still wouldn't have a way to actually view the persisted audit logs without looking in Redis or Postgres itself. So we'd probably want to add a UI in `FunWithFlags.UI` for viewing the audit logs related to a given key. This could happen separately from all of the other changes.

## In Conclusion

What do you think of this proposal? I think I understand enough about this project at this point to make these changes, but super open to feedback!

If you think this sounds like a reasonable new feature, I'd be happy to implement it. The only breaking changes I could think of is the additional callbacks on the `FunWithFlags.Store.Persistent` behavior, but at the very least it should warn people that have written custom persistence modules at compile time that they need to implement a new thing.

I'm confident that this can be built in a way that doesn't disrupt existing users (beyond those that have custom persistence) and allows a user to not configure audit logging and simply ignore this feature.

**Comment by tompave** [comment]
Hello!

Thank you for using the library and suggesting this feature. Especially, thank you for taking the time to write such a detailed proposal.

I appreciate the value of an audit log for when feature flags are toggled, but I wouldn't feel comfortable adding so many new responsibilities to the library. Audit logging is also quite difficult to do right and it's not something where one size fits all, so I am a bit resistant to the idea of shipping an opinionated implementation with FWF.

That doesn't mean that it can't be done with a third party library or with a bit of user code in the host application, though!

You seem to have spent some time thinking about the suggested design, that is, your `FunWithFlags.add_audit_log/4` function as an opinionated interface to persist some data. I can see why it would work for your use case, but what if other people would prefer to send the audit log to a log drain, a time-series DB, or to a queue of some kind? (e.g. Kafka, Rabbit, SQS, etc) The same thing applies to the UI. Even though the idea is good, I'm not convinced that this kind of opinionated functionality belongs into FWF, sorry.

So my suggestion is that this sort of thing could be isolated in a third party library or implemented in the host application.

Hopefully, the design of FWF would make it easy, and I can give you some pointers.


The first thing I'd like to point out is that you need to consider your trust model. Your proposal focuses on auditing interactions through `FWF.UI`, and it doesn't seem concerned with flags being toggled with code. The reason is that since `add_audit_log/4` is a separate function, there is no guarantee that will be invoked when toggling flags. Sure, if it were integrated in `FWF.UI` it could be assumed that the two calls always happen together. (Although you correctly point our the need for transactions.) That alone wouldn't be enough however. `FWF.UI` is a convenient web interface for human operators, but flags can also be toggled in application code. There is no guarantee that a developer won't insert a rogue line somewhere to toggle a flag, or more simply do it within a `iex -S mix` session, without also calling `FunWithFlags.add_audit_log/4`. So the first question would be: are you concerned about that?

If you're not, then good news. Adding audit logging to FWF UI is quite simple. Just add a plug.

For example, extending the [FWF.UI Readme](https://github.com/tompave/fun_with_flags_ui#security):

```elixir
defmodule MyApp.AuditLog.Plug do
  def init(opts), do: opts

  def call(conn, _opts) do
    # 1. Skip if it's an asset request.
    # 2. Get from the `conn` the flag and gate names and the action.
    # 3. Get the user, based on the logic of your app.
    # 3. Log what you need, e.g. write to the DB or to a queue.
    # 4. Maybe use `Plug.Conn.register_before_send/2` for some more logging?
    # 5. Done. Return the conn.
    conn
  end
end

# ...

pipeline :flag_auditing do
  plug MyApp.AuditLog.Plug
end

scope path: "/feature-flags" do
  pipe_through :mounted_and_protected_apps
  pipe_through :flag_auditing
  forward "/", FunWithFlags.UI.Router, namespace: "feature-flags"
end

```

You might argue that this doesn't give you the atomicity guarantees of a transaction, but I'd argue that it's better. This approach would allow you to have audit logs that follow the pattern: "user Foo is about to perform action Bar". Then with `Plug.Conn.register_before_send/2` you can look into the `conn` struct and add logs like "the  action was/was-not successful". You can even store unique IDs in the `conn` to ensure that the before and after logs are matched or end up together.

On top of that, you have full control over how the user is retrieved.

If, on the other hand, you also want to audit-log when flags are toggled _outside_ of `FWF.UI`, then you must put your audit-logging at a deeper level. That means it has to go directly into FWF.

A simple way to do it would be to add to FWF a way to register a callback that happens before/after/around the action of updating any persisted flag data. That would work, but I'd say it's not necessary. The same thing can be achieved with a [custom persistence adapter](https://github.com/tompave/fun_with_flags#custom-persistence-adapters). Just implement your own adapter that wraps one of the builtin adapters, and add your audit-logging code there:

```elxir
defmodule MyApp.FunWithFlags.PersistentStoreWithAuditLog do
  @behaviour FunWithFlags.Store.Persistent
  @base FunWithFlags.Store.Persistent.Ecto

  @impl true
  def put(flag_name, gate) do
    user = maybe_read_it_from_the_process_dictionary()
    MyApp.Auditing.add_audit_log(:toggle, flag_name, gate.type, user)
    @base.get(flag_name)
  end

  # No idea if this would work or if you need to expand them all!
  defdelegate everything_else, to: @base
end
```

Of course, a determined bad actor can still find a way to bypass that and write directly to the datastore, but at least that should cover the common cases!

**Comment by mbramson** [comment]
Thanks for the extremely detailed response! Like you pointed out, while a callback would be great... getting the user data there without using something gnarly like the process dictionary is a challenge. Wrapping it in a plug is probably the way to go.
